rdprail-shell/app-list.c

```diff
@@ -359,170 +368,125 @@ trim_command_exec(char *s)
-		/*      Need $XDG_CURRENT_DESKTOP keyword for WSL GUI environment */
-		if (weston_config_section_get_string(section,
-			 "OnlyShowIn", &s, NULL) == 0) {
-			shell_rdp_debug(shell, "desktop file: %s has OnlyShowIn %s\n", entry->file, s);
+	if (!g_key_file_has_group(key_file, GROUP_DESKTOP_ENTRY)) {
+		shell_rdp_debug(shell, "desktop file: %s is missing %s section\n",
+			entry->file, GROUP_DESKTOP_ENTRY);
+		goto exit;
+	}
+
+	if (g_key_file_get_boolean(key_file, GROUP_DESKTOP_ENTRY, "Hidden", NULL)) {
+		shell_rdp_debug(shell, "desktop file: %s is hidden\n",
+			entry->file);
+		goto exit;
+	}
+
+	s = g_key_file_get_string(key_file, GROUP_DESKTOP_ENTRY, "Type", NULL);
+	if (s) {
+		if (strcmp(s, "Application") != 0) {
+			shell_rdp_debug(shell, "desktop file: %s is not app (%s)\n", entry->file, s);
 			free(s);
-			return false; // terminal based app is not included.
+			goto exit;
 		}
-		if (app_list_config_section_get_string_by_language_id(section,
-			 "Name", &s, NULL, lang_id) == 0) {
-			if (shell->is_appid_with_distro_name) {
-				char *t;
-				size_t len = strlen(s);
-				/* 4 = ' ' + '(' + ')' + null */
-				len += (4 + shell->distroNameLength);
-				t = zalloc(len);
-				if (t) {
-					copy_string(t, len, s);
-					append_string(t, len, " (");
-					append_string(t, len, shell->distroName);
-					append_string(t, len, ")");
-					entry->name = t;
-					free(s);
-				} else {
-					entry->name = s;
-				}
-			} else {
-				entry->name = s;
-			}
-		} else {
-			/* name is required */
-			return false;
-		}
-		if (weston_config_section_get_string(section,
-			 "Exec", &s, NULL) == 0) {
-			entry->exec = trim_command_exec(s);
-		} else {
-			/* exec is required */
-			return false;
-		}
-		if (weston_config_section_get_string(section,
-			 "TryExec", &s, NULL) == 0) {
-			entry->try_exec = trim_command_exec(s);
-		}
-		if (weston_config_section_get_string(section,
-			 "Path", &s, NULL) == 0) {
-			entry->working_dir = s;
+		free(s);
+	}
+	if (g_key_file_get_boolean(key_file, GROUP_DESKTOP_ENTRY, "NoDisplay", NULL)) {
+		shell_rdp_debug(shell, "desktop file: %s has NoDisplay specified\n", entry->file);
+		goto exit; // NoDisplay app is not included.
+	}
+	if (g_key_file_get_boolean(key_file, GROUP_DESKTOP_ENTRY, "Terminal", NULL)) {
+		shell_rdp_debug(shell, "desktop file: %s is terminal based app\n", entry->file);
+		goto exit; // terminal based app is not included.
+	}
+	/*TODO: OnlyShowIn/NotShowIn support for WSL environment. */
+	/*      Need $XDG_CURRENT_DESKTOP keyword for WSL GUI environment */
+	s = g_key_file_get_string(key_file, GROUP_DESKTOP_ENTRY, "OnlyShowIn", NULL);
+	if (s) {
+		shell_rdp_debug(shell, "desktop file: %s has OnlyShowIn %s\n", entry->file, s);
+		free(s);
+		goto exit;
+	}
+	entry->name = g_key_file_get_locale_string(key_file, GROUP_DESKTOP_ENTRY, "Name", lang_id, NULL);
+	if (!entry->name) {
+		/* name is required */
+		goto exit;
+	} if (shell->is_appid_with_distro_name) {

```

Comment:
I'd like to see an else or a newline here. I guess preferably a newline since the previous block is already jumping elsewhere with a goto.

Reply:
Right, it looks I didn't press enter key strong enough...